modules/cas_cache/layer_cache_management.c

```diff
@@ -15,141 +15,367 @@ extern u32 seq_cut_off_mb;
 extern u32 use_io_scheduler;
 
 struct _cache_mngt_sync_context {
-	struct completion compl;
+	struct completion *compl;
 	int *result;
 };
 
+struct _cache_mngt_async_context {
+	struct completion compl;
+	atomic_t ref_count;
+	int result;
+};
+
+struct _cache_mngt_async_lock_context {
+	struct completion compl;
+	spinlock_t lock_sync;
+	int result;
+};
+
+/*
+ * Value used to mark async call as completed. Used when OCF call already
+ * finished, but complete function still has to be performed.
+ */
+#define ASYNC_CALL_FINISHED 1
+
 static void _cache_mngt_lock_complete(ocf_cache_t cache, void *priv, int error)
 {
-	struct _cache_mngt_sync_context *context = priv;
+	struct _cache_mngt_async_lock_context *context = priv;
+	bool interrupt = (context->result == -KCAS_ERR_WAITING_INTERRUPTED);
+
+	spin_lock(&context->lock_sync);
```

Comment:
in completion path you must use spin_lock_irqsave() to assure proper behavior in case completion is called with interrupts disabled (which is typical)